networkx/algorithms/swap.py

```diff
@@ -6,7 +6,130 @@
+        If nswap > max_tries, or
+        If there are fewer than 4 nodes in `G`
+    NetworkXAlgorithmError
+        If the number of swap attempts exceeds `max_tries` before `nswap` swaps are made
+
+    Notes
+    -----
+    Does not enforce any connectivity constraints.
+
+    The graph G is modified in place.
+
+    References
+    ----------
+    .. [1] Erdős, Péter L., et al. “A Simple Havel-Hakimi Type Algorithm to Realize
+           Graphical Degree Sequences of Directed Graphs.” ArXiv:0905.4913 [Math],
+           Jan. 2010. https://doi.org/10.48550/arXiv.0905.4913.
+    .. [2] “Combinatorics - Reaching All Possible Simple Directed Graphs with a given
+           Degree Sequence with 2-Edge Swaps.” Mathematics Stack Exchange,
+           https://math.stackexchange.com/questions/22272/. Accessed 30 May 2022.
+    """
+    if not G.is_directed():
+        raise nx.NetworkXError(
+            "directed_edge_swap() is only defined for directed graphs. Use double_edge_swap "
+            "instead"
+        )
+    if nswap > max_tries:
+        raise nx.NetworkXError("Number of swaps > number of tries allowed.")
+    if len(G) < 4:
+        raise nx.NetworkXError("Graph has less than four nodes.")
+
+    # Instead of choosing uniformly at random from a generated edge list,
+    # this algorithm chooses nonuniformly from the set of nodes with
+    # probability weighted by degree.
+    n = 0
+    swapcount = 0
+    keys, degrees = zip(*G.degree())  # keys, degree
+    cdf = nx.utils.cumulative_distribution(degrees)  # cdf of degree
+    discrete_sequence = nx.utils.discrete_sequence
+
+    while swapcount < nswap:
+        # choose source node index from discrete distribution
+        start_index = discrete_sequence(1, cdistribution=cdf, seed=seed)[0]
+        start = keys[start_index]
+
+        # If the given node doesn't have any out edges, then there isn't anything to swap
+        if G.out_degree(start) == 0:
+            n += 1
+            continue
+        _, second = seed.choice(list(G.out_edges(start)))
+
+        if G.out_degree(second) == 0:
+            n += 1
+            continue
+        _, third = seed.choice(list(G.out_edges(second)))
+
+        if G.out_degree(third) == 0:
+            n += 1
+            continue
+        _, fourth = seed.choice(list(G.out_edges(third)))
+
+        if start == second or second == third or third == fourth:
+            n += 1
+            continue
+
+        if (
+            (start, third) not in G.out_edges(start)
+            and (second, fourth) not in G.out_edges(second)
+            and (third, second) not in G.out_edges(third)
+        ):
+            # Swap nodes
+            G.add_edge(start, third)
+            G.add_edge(third, second)
+            G.add_edge(second, fourth)
+            G.remove_edge(start, second)
+            G.remove_edge(second, third)
+            G.remove_edge(third, fourth)
+            swapcount += 1
+
+        if n >= max_tries:
+            e = (
```

Comment:
Can we name `e` something else since we typically save e for an exception rather than a message?
We often use `msg.  But really, you can write the exception with it’s message all on one line and let `black` choose how to format it.